File: src/dataset/python/robot-simulator_exercism-ThomasZumsteg/openAI/ChatGPT4_robot_simulator_v2.py
# NORTH, EAST, SOUTH, WEST are directions a robot can face
NORTH, EAST, SOUTH, WEST = 0, 1, 2, 3

class Robot:
    """Robot is a simulation of a robot"""

    # Commands that the robot can be given
    _valid_commands = {
        'R': 'turn_right',
        'L': 'turn_left',
        'A': 'advance',
    }

    def __init__(self, bearing=NORTH, x=0, y=0):
        """__init__ creates the robot, defaults to facing north at the origin"""
        self.bearing = bearing
        self.coordinates = (x, y)

    def turn_right(self):
        """turn_right turns the robot to the right"""
        self.bearing = (self.bearing + 1) % 4

    def turn_left(self):
        """turn_left turns the robot to the left"""
        self.bearing = (self.bearing - 1) % 4

    def advance(self):
        """advance moves the robot one space forward in the direction it's facing"""
        x, y = self.coordinates
        moves = {
            NORTH: (x, y + 1),
            SOUTH: (x, y - 1),
            EAST: (x + 1, y),
            WEST: (x - 1, y),
        }
        self.coordinates = moves[self.bearing]

    def simulate(self, commands):
        """simulate gives the robot a set of commands"""
        for c in commands:
            getattr(self, Robot._valid_commands[c])()

File: src/dataset/javascript/robot-simulator_exercism-ThomasZumsteg/openAI/ChatGPT4_robot-simulator_v2.js
class Robot {
  constructor() {
    this.bearing = null;
    this.coordinates = [0, 0];
  }

  static directions = ['north', 'east', 'south', 'west'];
  static vectors = { north: [0, 1], east: [1, 0], south: [0, -1], west: [-1, 0] };
  static instructionsMap = { L: 'turnLeft', R: 'turnRight', A: 'advance' };

  orient(direction) {
    if (!Robot.directions.includes(direction)) throw "Invalid Robot Bearing";
    this.bearing = direction;
  }

  turnRight() {
    const currentIndex = Robot.directions.indexOf(this.bearing);
    this.bearing = Robot.directions[(currentIndex + 1) % 4];
  }

  turnLeft() {
    const currentIndex = Robot.directions.indexOf(this.bearing);
    this.bearing = Robot.directions[(currentIndex + 3) % 4];
  }

  at(x, y) {
    this.coordinates = [x, y];
  }

  advance() {
    const [dx, dy] = Robot.vectors[this.bearing];
    this.coordinates[0] += dx;
    this.coordinates[1] += dy;
  }

  instructions(instructionList) {
    return [...instructionList].map(instruction => Robot.instructionsMap[instruction]);
  }

  place({ direction, x, y }) {
    this.orient(direction);
    this.at(x, y);
  }

  evaluate(instructionList) {
    this.instructions(instructionList).forEach(action => this[action]());
  }
}

module.exports = Robot;
exports.directions = Robot.directions;